modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/FullPageId.java

```diff
@@ -0,0 +1,176 @@
+ *
+ * <p>The structure of a page ID is shown in the diagram below:
+ * <pre>
+ * +---------+-----------+------------+--------------------------+
+ * | 8 bits  |   8 bits  |  16 bits   |         32 bits          |
+ * +---------+-----------+------------+--------------------------+
+ * |  OFFSET |   FLAGS   |PARTITION ID|       PAGE INDEX         |
+ * +---------+-----------+------------+--------------------------+
+ * </pre>
+ *
+ * <p><h3>Page ID rotation</h3>
+ * There are scenarios when we reference one page (B) from within another page (A) by page ID. It is also
+ * possible that this first page (B) is concurrently reused for a different purpose. In this
+ * case we should have a mechanism to determine that the reference from page (A) to page (B) is no longer valid.
+ * This is ensured by page ID rotation - together with page's (B) ID we should write some value that is incremented
+ * each time a page is reused (page ID rotation). This ID should be verified after page read and a page
+ * should be discarded if full ID is different.
+ *
+ * <p>Effective page ID is page ID with zeroed bits used for page ID rotation.
+ */
+public class FullPageId {
+    /** Null page ID. */
+    public static final FullPageId NULL_PAGE = new FullPageId(-1, -1);
+
+    /** Page ID. */
+    private final long pageId;
+
+    /** Effective page ID. */
+    private final long effectivePageId;
+
+    /** Group ID. */
+    private final int groupId;
+
+    /**
+     * Constructor.
+     *
+     * @param pageId  Page ID.
+     * @param groupId Group ID.
+     */
+    public FullPageId(long pageId, int groupId) {
+        this.pageId = pageId;
+        this.groupId = groupId;
+
+        effectivePageId = PageIdUtils.effectivePageId(pageId);
+    }
+
+    /**
+     * Returns a page ID.
+     */
+    public long pageId() {
+        return pageId;
+    }
+
+    /**
+     * Returns an effective page ID.
+     */
+    public long effectivePageId() {
+        return effectivePageId;
+    }
+
+    /**
+     * Returns a group ID.
+     */
+    public int groupId() {
+        return groupId;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+
+        if (!(o instanceof FullPageId)) {
+            return false;
+        }
+
+        FullPageId that = (FullPageId) o;
+
+        return effectivePageId == that.effectivePageId && groupId == that.groupId;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public int hashCode() {
+        return hashCode0(groupId, effectivePageId);
+    }
+
+    /**
+     * Returns hash code of the ID.
+     *
+     * @param groupId Group ID.
+     * @param pageId  Page ID.
+     * @return Hash code.
+     */
+    public static int hashCode(int groupId, long pageId) {
+        long effectiveId = PageIdUtils.effectivePageId(pageId);
+
+        return IgniteUtils.hash(hashCode0(groupId, effectiveId));
```

Comment:
I wonder why it's hashed second time. `IgniteUtils#hash()` says that it's protecting from poor hash functions. Is `hashCode0()` a poor hash function? If so, why isn't `IgniteUtils#hash()` applied in the parameterless `hashCode()` overload?

Reply:
I can't answer yet, we'll take a look at these hashes once code that uses them is ported